XpressAI/xircuits
# 💥Dropped link to add node and automatically connect its link via components panel

# Description

This will enable to add node via context menu. It's also possible to automatically connect its link by dropping a link to canvas.

The components panel will only display what component it can linked but not smart enough to know 'Can it connect?' as there is **no type check** when the link is connecting.

There are 3 condition to display 3 different components list which are:

1. Right-clicking canvas will show all component that can be added.
2. Dropping link from ▶ either `inPort` or `Outport` will show all but General Components. (Can be added and linked)
3. Dropping link from any parameter `inPort`(beside ▶) will only display General Components. (Can be added and linked)

**Take note:**
**Dropping link from any parameter `OutPort`(beside ▶) will not display anything as it's hard to capture what it can link to.**

As this component panel can be activated by right-clicking, it make the previous Jupyterlab's context menu obsolete. With that, new custom context menu was created with fewer command mostly for nodes. It can be activated by **ctrl+left-click**.

Also, refactored and removed all the general components except literal and hyperparameter.

## Pull Request Type


## Type of Change


# Tests

1. Try right-clicking canvas.
   1. A panel showing components will opened. (The previous context menu was removed)
   2. Choose any node to add it.
2. **Try dropping link** based on the condition mentioned above and see it's the same as mentioned
   1. Link from ▶ either from inPort or OutPort.
   2. Link from any parameter `inPort`(beside ▶).
3. Also, when choosing the node, does the **link automatically connect**?
4. When component panel is displayed, try using the **search bar** and click the display node.
   1. Try dropping different link and search again.
   2. It should only show the allowable components.
5. Try opening the **new context menu** by **ctrl+left-click**.
   1. Try each option and does it work as expected.
6. Try **ctrl+left-click** any node excpet general components at the components panel
   1. It'll open its python script

## Tested on?


# Notes

Currently, there is **no type check** when connecting the link. Also, feels like some of the files in `/src/component` folder need to move to different folder.
